src/simplewallet/simplewallet.cpp

```diff
@@ -8354,3 +8581,1025 @@ int main(int argc, char* argv[])
+  for (size_t i = 0; i < members.size(); ++i)
+  {
+    const mms::coalition_member &member = members[i];
+    std::string label = member.label.empty() ? tr("<not set>") : member.label;
+    std::string monero_address;
+    if (member.monero_address_known)
+    {
+      monero_address = get_account_address_as_str(m_wallet->nettype(), false, member.monero_address);
+    }
+    else
+    {
+      monero_address = tr("<not set>");
+    }
+    std::string transport_address = member.transport_address.empty() ? tr("<not set>") : member.transport_address;
+    success_msg_writer() << boost::format("%2s %-20s %-s") % (i + 1) % label % transport_address;
+    success_msg_writer() << boost::format("%2s %-20s %-s") % "" % member.auto_config_token % monero_address;
+    success_msg_writer() << "";
+  }
+}
+
+void simple_wallet::add_member_config_messages()
+{
+  mms::message_store& ms = m_wallet->get_message_store();
+  std::string member_config;
+  ms.get_member_config(member_config);
+
+  const std::vector<mms::coalition_member> members = ms.get_all_members();
+  mms::multisig_wallet_state state = get_multisig_wallet_state();
+  uint32_t coalition_size = ms.get_coalition_size();
+  for (size_t i = 1 /* without me */; i < coalition_size; ++i)
+  {
+    ms.add_message(state, i, mms::message_type::member_config, mms::message_direction::out, member_config);
+  }
+}
+
+void simple_wallet::show_message(const mms::message &m)
+{
+  mms::message_store& ms = m_wallet->get_message_store();
+  const mms::coalition_member &member = ms.get_member(m.member_index);
+  bool display_content;
+  switch (m.type)
+  {
+  case mms::message_type::key_set:
+  case mms::message_type::additional_key_set:
+  case mms::message_type::note:
+    display_content = true;
+    break;
+  default:
+    display_content = false;
+  }
+  uint64_t now = time(NULL);
+  success_msg_writer() << "";
+  success_msg_writer() << tr("Message ") << m.id;
+  success_msg_writer() << tr("In/out: ") << ms.message_direction_to_string(m.direction);
+  success_msg_writer() << tr("Type: ") << ms.message_type_to_string(m.type);
+  success_msg_writer() << tr("State: ") << boost::format(tr("%s since %s, %s ago")) %
+          ms.message_state_to_string(m.state) % get_human_readable_timestamp(m.modified) % get_human_readable_timespan(std::chrono::seconds(now - m.modified));
+  if (m.sent == 0)
+  {
+    success_msg_writer() << tr("Sent: Never");
+  }
+  else
+  {
+    success_msg_writer() << boost::format(tr("Sent: %s, %s ago")) %
+            get_human_readable_timestamp(m.sent) % get_human_readable_timespan(std::chrono::seconds(now - m.sent));
+  }
+  success_msg_writer() << tr("Member: ") << ms.member_to_string(member, 100);
+  success_msg_writer() << tr("Content size: ") << m.content.length() << tr(" bytes");
+  success_msg_writer() << tr("Content: ") << (display_content ? m.content : tr("(binary data)"));
```

Comment:
Yes, and maybe that's an important security consideration that I totally overlooked.

In case of `note` messages, where coalition members can send short textual notes to each other to inform about something, it's even the whole point that the `mms show` command displays "attacker controlled data". (The idea: It's more comfortable and faster if people can message each other directly from within the CLI wallet.)

There is *some* protection now because regular `note` message construction goes through the CLI command parameter parser which won't allow arbitrary nonsense as note text, but of course without much trouble an attacker could doctor this and send whatever they like.

Is there already a precedent somewhere in the codebase how to handle something like this? If not, how about filtering out control characters (code < 32) when displaying and limiting notes to a sane length of say 500 characters?

Reply:
Misparsed HTTP data was logged. That's all I can think of atm.

Reply:
I implemented a "sanitize" method that strips out any controls, and < > characters that could be used to form any XML or HTML tags, and restricts the display to the first 1000 characters. I think that should do it, given the fact that not *anybody* can send me messages, but only one of the other authorized signers.